### PR TITLE
enh(contacts): show/hide addressbooks for all

### LINKED ISSUE #479: Add new option to hide an address book's contacts without disabling the address book
This issue results from dialog in [Enable/disable address books #412](https://github.com/nextcloud/contacts/pull/412).

The new option already implemented to disable an address book seems to do that on such a low level that it is then disabled per se - for all purposes. For this reason, it is not possible to disable an address book which has been shared with you rather than belonging to you.

In many use cases - team work, small to medium businesses, large contact archives used only seldom - we need a different function: to HIDE the contacts of a particular address book from the VIEW of the logged-in user (available also when not the owner). This is especially true as the historical design of the contacts app is to load all the visible contacts to the front end and then do all the nifty things like selection by category in JS. Which is great with up to (nowadays) about 1000 contacts, but becomes sluggish to unbearable the more you go over the top of that limit.

I would personally prefer that hidden contacts would still be searchable when I type in part of a name or email in the Contacts app search box. And I would certainly want that contacts hidden in the Contacts app (but not disabled) be active for the mail app, e.g. to make sure that mails from contacts known to me are not put into the spam folder.

Finally, in the settings area, if the little "dotdotdot" dropdown for each address book gets a further option "hide/show", then we should really try to explain the difference between Disable and Hide right there where the user has that choice - are tooltips possible there?

Some remote part of me is still dreaming about a backend-based variant of this app - termed by yours truly "Big Contacts" - which would allow working with many thousands of contacts, with the frontend fetching them piece by piece from the backend via traditional lazy loading / infinite scrolling, [see this issue](https://help.nextcloud.com/t/handling-a-large-number-1000-of-contacts-in-the-web-ui/22064).

Be that as it may, the abilty to hide an address book would be a great step on - so thanks John, thanks Henni, thanks Alexander!

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/55015861-add-new-option-to-hide-an-address-book-s-contacts-without-disabling-the-address-book?utm_campaign=plugin&utm_content=tracker%2F46751899&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F46751899&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by JohannesGGE**
* Fixes https://github.com/nextcloud/contacts/issues/479

## TODO

- [ ] ...

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by JohannesGGE** [comment]
Due to the different implementation of the recently contacted addressbook, the call for changing the enabled property can't be executed, and an "addressbook is immutable" is thrown in `apps/contactsinteraction/lib/AddressBook.php::propPatch()`.

Other addressbooks call `apps/dav/lib/CardDAV/AddressBook.php::propPatch()`

As far as I can tell the only needed code is the DB update of `apps/dav/lib/DAV/CustomPropertiesBackend.php::updateProperties()` to have a working enable/disable for the recently contacted addressbook.

**Comment by JohannesGGE** [comment]
Recently contacted address book will be tackled in another issue.

File: apps/dav/lib/CardDAV/AddressBook.php [hunk 1]
```diff
@@ -221,9 +226,6 @@ public function delete() {
 	}
 
 	public function propPatch(PropPatch $propPatch) {
-		if (isset($this->addressBookInfo['{http://owncloud.org/ns}owner-principal'])) {
```
Comment (ChristophWurst) [inline apps/dav/lib/CardDAV/AddressBook.php:224]: why is this dropped?

File: apps/dav/lib/CardDAV/AddressBook.php [hunk 2]
```diff
@@ -116,7 +116,12 @@ public function getACL() {
 				'privilege' => '{DAV:}write',
 				'principal' => $this->getOwner(),
 				'protected' => true,
-			]
+			],
+            [
+                'privilege' => '{DAV:}write-properties',
```
Comment (ChristophWurst) [inline apps/dav/lib/CardDAV/AddressBook.php:121]: so this means users are now allowed to update properties of all address books, right? are there any property changes we might not want? e.g. rename.

File: apps/dav/lib/CardDAV/AddressBook.php (hunk 1 above)
Comment (JohannesGGE) [reply apps/dav/lib/CardDAV/AddressBook.php:224]: owner-principal seems to be set if the address book is shared, system or recently contacted. Own address books don't have this prop.

File: apps/dav/lib/CardDAV/AddressBook.php (hunk 2 above)
Comment (JohannesGGE) [reply apps/dav/lib/CardDAV/AddressBook.php:121]: just tested it with renaming, and it turns out the change of the name is not stored as a user prop but changes the real name, so the owner sees the change as well. I'll have a look in the code again to check where the distinction between the props is made.

Comment (JohannesGGE) [reply apps/dav/lib/CardDAV/AddressBook.php:121]: @tcitworld As far as I understand it, the propPatch() in CustomPropertiesBackend.php is just called if the prop is a user prop. So the distinction is done somewhere before

**Comment by JohannesGGE** [comment]
The addressbook is now checked for sharing. If it is shared the props will be saved as userProps. So even if the addressbook is renamed by user that is not the owner (not possible in the contacts app yet), the owners addressbook name isn't changed. For the case the addressbook has a userProp name, this name is shown now.

File: apps/dav/lib/DAV/CustomPropertiesBackend.php [hunk 3]
```diff
@@ -176,6 +176,21 @@ public function propFind($path, PropFind $propFind) {
...
+				'{DAV:}displayname',
+			];
+
+			foreach ($customPropertiesForShares as $customPropertyForShares) {
+				if (in_array($customPropertyForShares, $allRequestedProps)) {
+					$requestedProps[] = $customPropertyForShares;
+				}
+			}
+		}
+
```
Comment (tcitworld) [inline apps/dav/lib/DAV/CustomPropertiesBackend.php:193]: **suggestion:** I think there's a way to factorize this block with the block above in a single function which takes calendars and addressbooks specifics as parameters.

File: apps/dav/lib/DAV/CustomPropertiesBackend.php [hunk 4]
```diff
@@ -176,6 +176,21 @@ public function propFind($path, PropFind $propFind) {
 			}
 		}
 
+		// substr of addressbooks/ => path is inside the CardDAV component
+		// three '/' => this a addressbook (no addressbook-home nor contact object)
+		if (substr($path, 0, 13) === 'addressbooks/' && substr_count($path, '/') === 3) {
```
Comment (ChristophWurst) [inline apps/dav/lib/DAV/CustomPropertiesBackend.php:181]: server is PHP8.0+ so you can use https://www.php.net/manual/en/function.str-starts-with.php

File: apps/dav/lib/DAV/CustomPropertiesBackend.php [hunk 5]
```diff
@@ -176,6 +176,21 @@ public function propFind($path, PropFind $propFind) {
...
+			$allRequestedProps = $propFind->getRequestedProperties();
+			$customPropertiesForShares = [
+				'{DAV:}displayname',
+			];
+
+			foreach ($customPropertiesForShares as $customPropertyForShares) {
+				if (in_array($customPropertyForShares, $allRequestedProps)) {
```
Comment (ChristophWurst) [inline apps/dav/lib/DAV/CustomPropertiesBackend.php:188]:
```suggestion
				if (in_array($customPropertyForShares, $allRequestedProps, true)) {
```

for strict comparison

**Review by miaulalala — APPROVED** [review]

**Comment by kesselb** [comment]
nitpick: enh(contacts) should be feat(contacts) according to https://www.conventionalcommits.org/en/v1.0.0/

File: apps/dav/lib/CardDAV/AddressBook.php (hunk 1 above)
Comment (kesselb) [reply apps/dav/lib/CardDAV/AddressBook.php:224]: The change also surprised me. 

Apparently, this was added before the CustomPropertiesBackend: https://github.com/nextcloud/server/commit/95e218b00c9526bf3cba8ce719b09b1ea9d49ba9

If we throw forbidden here, the CustomPropertiesBackend::propPatch is not executed?

**Review by kesselb — APPROVED** [review]
Well done :tada:

**Pull request merged**